drivers: i2c_ll_stm32_v2: add support for second target address

aurel32 opened the pull request:
The Zephyr API supports multiple I2C targets addresses, and the STM32
I2C v2 implementation allows to define up to 2 targets addresses.
    
This patch adds support for a second I2C target address. It adds a new
config entry in the i2c_stm32_data structure, and uses the fact that
both addresses can be enabled and disabled independently. In the
interrupt, the target being addressed is determined using the address
match code from the interrupt status register.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
drivers: i2c_ll_stm32_v2: add support for second target address

The Zephyr API supports multiple I2C targets addresses, and the STM32
I2C v2 implementation allows to define up to 2 targets addresses.

This patch adds support for a second I2C target address. It adds a new
config entry in the i2c_stm32_data structure, and uses the fact that
both addresses can be enabled and disabled independently. In the
interrupt, the target being addressed is determined using the address
match code from the interrupt status register.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

gmarull commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -118,21 +118,34 @@ static void stm32_i2c_slave_event(const struct device *dev)
+	} else if (slave_address == data->slave2_cfg->address) {
+		slave_cfg = data->slave2_cfg;
+	} else {
+		/* This should not happen, just return */
```
maybe assert here?

aurel32 replied:
I have been looking at the code again, I think this could actually happen during a master transition in case a slave is attached. In that case just exiting is the correct thing to do I'll update the comment accordingly.

aurel32 replied:
Sorry, I am actually wrong, this code is only executed when `!data->master_active`, so this indeed should not apply. Is it fine adding an assert in interrupt code?

aurel32 replied:
I have found that `__ASSERT_NO_MSG` is used in other ISR code, so that should be fine. I have just pushed a new version using that.

gmarull commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -118,21 +118,34 @@ static void stm32_i2c_slave_event(const struct device *dev)
+	uint16_t slave_address;
+
+	/* Choose the right slave from the address match code */
+	slave_address = LL_I2C_GetAddressMatchCode(i2c) >> 1;
```
note: according to docs, `LL_I2C_GetAddressMatchCode` range of return values is `@retval Value between Min_Data=0x00 and Max_Data=0x3F`, so maybe use `uint8_t` here?

aurel32 replied:
The documentation is definitely wrong, it should be `Max_data=0xff`. That said your comment is still valid, I'll update the code.

gmarull commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -207,17 +220,26 @@ int i2c_stm32_slave_register(const struct device *dev,
+		data->slave_cfg = config;
 
-	data->slave_attached = true;
+		LL_I2C_SetOwnAddress1(i2c, config->address << 1,
```
```suggestion
		LL_I2C_SetOwnAddress1(i2c, config->address << 1U,
```

aurel32 replied:
This comes from the original code, but your point is valid, I'll fix it.

gmarull commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -207,17 +220,26 @@ int i2c_stm32_slave_register(const struct device *dev,
+	} else {
+		data->slave2_cfg = config;
+
+		LL_I2C_SetOwnAddress2(i2c, config->address << 1,
```
```suggestion
		LL_I2C_SetOwnAddress2(i2c, config->address << 1U,
```

aurel32 replied:
Ditto

erwango reviewed (dismissed):
To be synchronized with https://github.com/zephyrproject-rtos/zephyr/pull/43166

aurel32 commented:
> To be synchronized with #43166

Given that #43166 has been merged, I have just rebased this PR, and updated the commit description to use the new terminology.
